tests/cpp_unit.cpp

```diff
@@ -828,6 +831,97 @@ DEFINE_TEST(test_cpp_remove_range_64) {
     }
 }
 
+std::pair<doublechecked::Roaring64Map, doublechecked::Roaring64Map>
+    makeTwoBigRoaring64Maps() {
+    // Insert a large number of pseudorandom numbers into two sets.
+    const uint32_t randomSeed = 0xdeadbeef;
+    const size_t numValues = 1000000;  // 1 million
+
+    doublechecked::Roaring64Map roaring1;
+    doublechecked::Roaring64Map roaring2;
+
+    std::default_random_engine engine(randomSeed);
+    std::uniform_int_distribution<uint64_t> rng;
+
+    for (size_t i = 0; i < numValues; ++i) {
+        auto value = rng(engine);
+        auto choice = rng(engine) % 4;
+        switch (choice) {
+            case 0: {
+                // Value is added only to set 1.
+                roaring1.add(value);
+                break;
+            }
+
+            case 1: {
+                // Value is added only to set 2.
+                roaring2.add(value);
+                break;
+            }
+
+            case 2: {
+                // Value is added to both sets.
+                roaring1.add(value);
+                roaring2.add(value);
+                break;
+            }
+
+            case 3: {
+                // Value is added to set 1, and a slightly different value
+                // is added to set 2. This makes it likely that they are in
+                // the same "outer" bin, but at a different "inner" position.
+                roaring1.add(value);
+                roaring2.add(value + 1);
```

Comment:
Kindly allow me to defend the approach a little bit, and then you can decide ultimately what to do about it. The `make_two_big_roaring64_maps` generator populates the bitsets with deterministic pseudorandom  keys which fall into four outer buckets:

A: keys that are likely only in set 1
B: keys that are likely only in set 2
C: keys that are in both sets
D: keys that are likely in the same outer bucket but different inner positions

The value of the keys don't *really* matter that much, and the fact that they're spread all over the `uint64_t` space is basically fine in my opinion. I'm not expecting them to ever collide (but it doesn't matter if they do). It's much more interesting to observe which of cases A-D they fall into.

I say "likely" because due to random effects we might get lucky or unlucky with duplicate keys but this doesn't really matter. I don't dispute your math in the above comment, but hitting the edge case neither causes anything interesting to happen nor breaks anything. Put another way, in case D our *intention* is to create two different keys that happen to belong to the same outer bucket but in different inner positions. Every once in a great while, if we happen to hit your edge case, in fact those keys end up in *different* outer buckets and different positions. So every once in a while we thought we were making a case D but we ended up making a case A plus case B. But it's all random anyway, so we don't care and getting lucky or unlucky doesn't break anything. There are lots of other ways to get lucky or unlucky (we might have a case A followed by the exact same key for case B and so we've "really" created a case C but again it's sort of like, that's the luck of the draw, no harm no foul).

The reason I implemented pattern A-D above is that it creates interesting cases for union, intersection, difference, and XOR. For example, given a million keys we would expect:
set 1 to have approximately 750K keys (250K each from cases A, C, and D)
set 2 to have approximately 750K keys (250K each from cases B, C, and D)
`(set1 & set2)` to have approximately 250K keys (from case C)
`(set1 | set2)` to have approximately 1.25M keys (250K each from cases A-C and 500K from case D)
`(set1 - set2)` to have approximately 500K keys (250K each from case A and case D)
`(set1 ^ set2)` to have approximately 1M keys (250K each from cases A and B, and 500K from case D)

When we add print statements that print these cardinalities, we see numbers that closely match these predictions:
set 1 size 750787
set 2 size 749667
intersection 250230
union 1250224
difference 500557
XOR 999994

So we have both the exact comparison from `does_std_set_match_roaring()` and also the aggregate statistical expectation which provides a little comfort that things are working as intended. Note I do not have print statements in the above code but I could print statements and/or assertions that say e.g. that union should result in exactly 1250224 elements.

Anyway, that's the rationale. I'm happy to proceed however you think makes sense. I do like your suggestion of iterating/refining the unit tests in a separate PR.

Reply:
Your rationale makes sense to me. I had actually misinterpreted the comment in the test to mean two adjacent outer positions rather than the same outer position. I think we should still add tests focused on specific scenarios, but let's leave that for later.